compiler/rustc_mir_build/src/check_unsafety.rs

```diff
@@ -246,11 +248,135 @@ impl<'a, 'tcx> Visitor<'a, 'tcx> for UnsafetyVisitor<'a, 'tcx> {
                 // Unsafe blocks can be used in closures, make sure to take it into account
                 self.safety_context = closure_visitor.safety_context;
             }
+            ExprKind::Borrow { borrow_kind, arg } => {
+                let arg = match self.thir[arg].kind {
+                    ExprKind::Scope { value, .. } => value,
+                    _ => arg,
+                };
+                if let ExprKind::Field { lhs, .. } = self.thir[arg].kind {
```

Comment:
This is incorrect and doesn't catch cases where e.g. the layout-constrained field is deeper:

```rs
#[rustc_layout_scalar_valid_range_start(1)]
struct NonZero<T>(T);

let mut foo = unsafe { NonZero((1,)) };
let a = &mut foo.0.0;
```

In fact, a place can contain all of these kinds of expressions: https://github.com/rust-lang/rust/blob/d34a3a401b4e44f289a4d5bf53da83367cbb6aa7/compiler/rustc_mir_build/src/build/expr/category.rs#L38-L44

So as long as we see these, we're still in the same place and we need to continue looking for layout-constrained fields.

Reply:
You are right; I have attempted to fix this now (your example correctly leads to an error in my implementation now). However, I think we have to stop at `Deref` expressions, which is also what the current MIR implementation seems to do, unless I am misunderstanding something:
https://github.com/rust-lang/rust/blob/308fc2322bf00b5dc12454489679de1420320f56/compiler/rustc_mir/src/transform/check_unsafety.rs#L344-L346

Reply:
Now it's too restrictive:

```rs
let foo = unsafe { NonZero((1,)) };
&mut { foo.0 }.0;
```

This compiles under MIR unsafeck because the block doesn't belong to the place and creates a new value by copying the content `(1,)`, which doesn't belong to `foo` anymore so the operation is safe. To fix this I'd expect you just need to also stop on any expression that's not in `Category::Place` (the list above).